src/compiler/utilities.ts

```diff
@@ -215,8 +215,37 @@ namespace ts {
 
     // Gets the nearest enclosing block scope container that has the provided node
     // as a descendant, that is not the provided node.
-    export function getEnclosingBlockScopeContainer(node: Node): Node {
-        let current = node.parent;
+    export function getEnclosingBlockScopeContainer(node: Node): Node;
+
+    /**
+      * Gets the nearest enclosing block scope container that has the provided node
+      * as a descendant, that is not the provided node.
+      * @param node The starting node
+      * @param getAncestorOrSelf A callback used to get the ancestor of the starting node, used 
+      * only when traversing the ancestors of the current node in the emitter.
+      * @param offset The offset in the node stack used to get the ancestor of the current node
+      * from the emitter's node stack.
+      */
+    export function getEnclosingBlockScopeContainer(node: Node, getAncestorOrSelf: (offset: number) => Node, offset: number): Node;

```

Comment:
Why is `getAncestorOrSelf()` a parameter? Why don't you just define a non-exported helper in the module and call it all over the place? Then we can implement it as parent-pointer traversal for now, and update it later everywhere.

Or do we actually want `getAncestorOrSelf()` to perform different functions in different parts of the code?